File: backend/app/api/pets.py
"""M4 灵宠 HTTP 路由。"""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.deps import (
    get_current_user,
    get_pet_duel_service,
    get_pet_explore_service,
    get_pet_hatch_service,
    get_pet_service,
    get_play_gate,
)
from app.db.models import User
from app.schemas.common import AppError, success
from app.schemas.pets import (
    PetAffixRerollValueRequest,
    PetCaptureTestRequest,
    PetDuelAutoRequest,
    PetDuelNpcStartRequest,
    PetDuelTurnRequest,
    PetExploreAutoRequest,
    PetExploreCaptureRequest,
    PetExploreEncounterRequest,
    PetFeedRequest,
    PetHatchStartRequest,
    PetPatchRequest,
    PetSectRerollTypeRequest,
    PetSkillLearnBookRequest,
    PetSkillLearnPoolRequest,
    PetSkillsEquipRequest,
)
from app.services.pet_duel_service import PetDuelService
from app.services.pet_explore_service import PetExploreService
from app.services.pet_hatch_service import PetHatchService
from app.services.pet_service import PetService, spirit_beast_sect_enabled
from app.services.play_gate import PlayGate
from app.services.realm_config import get_game_config

router = APIRouter(prefix="/pets", tags=["pets"])


@router.get("", response_model=None)
async def list_pets(
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """灵宠列表。"""
    character = await gate.require_character(current_user)
    pets = await service.list_pets(character)
    cfg = get_game_config().pets
    return success({"pets": pets, "hold_cap": cfg.hold_cap})


@router.get("/catalog", response_model=None)
async def pets_catalog(
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    灵宠图鉴：注册表投影 + seen/caught。

    加 YAML/后台物种后无需改代码即可出现。
    """
    character = await gate.require_character(current_user)
    data = await service.catalog(character)
    return success(data)


@router.get("/hatch", response_model=None)
async def pet_hatch_list(
    service: PetHatchService = Depends(get_pet_hatch_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """孵化面板：蛋目录 + 会话（N5）。"""
    character = await gate.require_character(current_user)
    data = await service.list_state(character)
    return success(data)


@router.post("/hatch/start", response_model=None)
async def pet_hatch_start(
    payload: PetHatchStartRequest,
    service: PetHatchService = Depends(get_pet_hatch_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """消耗蛋开启孵化会话。"""
    character = await gate.require_character(current_user)
    data = await service.start(character, egg_item_id=payload.egg_item_id)
    return success(data)


@router.post("/hatch/{job_id}/claim", response_model=None)
async def pet_hatch_claim(
    job_id: int,
    service: PetHatchService = Depends(get_pet_hatch_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """领取孵化完成的灵宠入园。"""
    character = await gate.require_character(current_user)
    data = await service.claim(character, job_id)
    return success(data)


@router.get("/sect/affix/type-reroll/preview", response_model=None)
async def pet_sect_type_reroll_preview(
    pet_id: int | None = None,
    slot_index: int | None = None,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    灵兽宗改词条类型预览（PET-D06）。

    返回费用公式；可选 ``pet_id`` + ``slot_index`` 给出单槽报价。
    """
    character = await gate.require_character(current_user)
    data = await service.type_reroll_preview_quote(
        character,
        pet_id=pet_id,
        slot_index=slot_index,
    )
    return success(data)


@router.get("/sect/affix/type-reroll/status", response_model=None)
async def pet_sect_type_reroll_status(
    pet_id: int,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """各槽改类型次数与下次费用（PET-D06）。"""
    character = await gate.require_character(current_user)
    data = await service.type_reroll_status(character, pet_id)
    return success(data)


@router.post("/sect/affix/reroll-type", response_model=None)
async def pet_sect_reroll_type(
    payload: PetSectRerollTypeRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    改词条类型：重 roll 指定槽类型/品级/数值并扣灵石。

    Raises:
        AppError: 50110 设施未开放 / 40061 槽不可改 / 40012 灵石不足。
    """
    _ = payload.idempotency_key  # 契约预留；本期不落幂等表
    if not spirit_beast_sect_enabled():
        raise AppError(
            code=50110,
            message="灵兽宗设施未开放（后台 sects.facilities.spirit_beast_sect）",
            http_status=501,
        )
    character = await gate.require_character(current_user)
    data = await service.reroll_affix_type(
        character,
        payload.pet_id,
        slot_index=payload.slot_index,
    )
    return success(data)


@router.post("/capture_test", response_model=None)
async def capture_test_pet(
    payload: PetCaptureTestRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """测试捕获样本宠。"""
    character = await gate.require_character(current_user)
    data = await service.capture_test(character, species_id=payload.species_id)
    return success(data)


@router.get("/explore/preview", response_model=None)
async def pet_explore_preview(
    region_id: str = "default",
    service: PetExploreService = Depends(get_pet_explore_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """野外遭遇池预览（M4-D04c）。"""
    character = await gate.require_character(current_user)
    data = await service.preview(character, region_id=region_id)
    return success(data)


@router.post("/explore/encounter", response_model=None)
async def pet_explore_encounter(
    payload: PetExploreEncounterRequest,
    service: PetExploreService = Depends(get_pet_explore_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """掷一次野外遭遇（区×时×天）。"""
    character = await gate.require_character(current_user)
    data = await service.encounter(
        character,
        region_id=payload.region_id,
        seed=payload.seed,
    )
    return success(data)


@router.post("/explore/capture", response_model=None)
async def pet_explore_capture(
    payload: PetExploreCaptureRequest,
    service: PetExploreService = Depends(get_pet_explore_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """野外捕获（诱灵草 + 全因子骰；成功 wild_capture 入园）。"""
    character = await gate.require_character(current_user)
    data = await service.capture(
        character,
        encounter_id=payload.encounter_id,
        seed=payload.seed,
    )
    return success(data)


@router.post("/explore/auto", response_model=None)
async def pet_explore_auto(
    payload: PetExploreAutoRequest,
    service: PetExploreService = Depends(get_pet_explore_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """自动探索捕：遇可捕灵兽则尝试一次。"""
    character = await gate.require_character(current_user)
    data = await service.auto_capture(
        character,
        region_id=payload.region_id,
        seed=payload.seed,
    )
    return success(data)


@router.post("/duel/npc/start", response_model=None)
async def pet_duel_npc_start(
    payload: PetDuelNpcStartRequest,
    service: PetDuelService = Depends(get_pet_duel_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """开启灵宠 vs NPC 回合制对战（PET-D05）。"""
    character = await gate.require_character(current_user)
    data = await service.start_npc(
        character,
        pet_id=payload.pet_id,
        npc_id=payload.npc_id,
        seed=payload.seed,
    )
    return success(data)


@router.post("/duel/npc/auto", response_model=None)
async def pet_duel_npc_auto(
    payload: PetDuelAutoRequest,
    service: PetDuelService = Depends(get_pet_duel_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """一键自动打完 vs NPC（seed 可复现）。"""
    character = await gate.require_character(current_user)
    data = await service.auto_npc(
        character,
        pet_id=payload.pet_id,
        npc_id=payload.npc_id,
        seed=payload.seed,
    )
    return success(data)


@router.get("/duel/{duel_id}", response_model=None)
async def pet_duel_get(
    duel_id: str,
    service: PetDuelService = Depends(get_pet_duel_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """读取对战快照。"""
    character = await gate.require_character(current_user)
    data = await service.get_duel(character, duel_id)
    return success(data)


@router.post("/duel/{duel_id}/turn", response_model=None)
async def pet_duel_turn(
    duel_id: str,
    payload: PetDuelTurnRequest,
    service: PetDuelService = Depends(get_pet_duel_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """提交选招并结算一回合。"""
    character = await gate.require_character(current_user)
    data = await service.turn(character, duel_id, skill_id=payload.skill_id)
    return success(data)


@router.post("/{pet_id}/upgrade", response_model=None)
async def upgrade_pet(
    pet_id: int,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """灵宠升级占位。"""
    character = await gate.require_character(current_user)
    data = await service.upgrade(character, pet_id)
    return success(data)


@router.post("/{pet_id}/grade-up", response_model=None)
async def grade_up_pet(
    pet_id: int,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    升阶：扣灵石、品阶+1、追加 1 条随机词条（PET-D01）。
    """
    character = await gate.require_character(current_user)
    data = await service.grade_up(character, pet_id)
    return success(data)


@router.post("/{pet_id}/feed", response_model=None)
async def feed_pet(
    pet_id: int,
    payload: PetFeedRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """丹药喂养：扣兽丹、涨面板；超上限拒绝（PET-D04）。"""
    character = await gate.require_character(current_user)
    data = await service.feed(
        character,
        pet_id,
        item_id=payload.item_id,
        quantity=payload.quantity,
    )
    return success(data)


@router.post("/{pet_id}/affix/reroll-value", response_model=None)
async def reroll_pet_affix_value(
    pet_id: int,
    payload: PetAffixRerollValueRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    数值-only 洗炼：不改类型与品级（PET-D01）。
    """
    character = await gate.require_character(current_user)
    data = await service.reroll_affix_value(
        character,
        pet_id,
        slot_index=payload.slot_index,
    )
    return success(data)


@router.post("/{pet_id}/skills/equip", response_model=None)
async def equip_pet_skills(
    pet_id: int,
    payload: PetSkillsEquipRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """装备最多 4 个已学主动技能（PET-D02）。"""
    character = await gate.require_character(current_user)
    data = await service.equip_skills(character, pet_id, equipped=list(payload.equipped))
    return success(data)


@router.post("/{pet_id}/skills/learn", response_model=None)
async def learn_pet_skill_from_pool(
    pet_id: int,
    payload: PetSkillLearnPoolRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """从物种技能池领悟技能（无等级门槛占位）。"""
    character = await gate.require_character(current_user)
    data = await service.learn_skill_from_pool(
        character,
        pet_id,
        skill_id=payload.skill_id,
    )
    return success(data)


@router.post("/{pet_id}/skills/learn_book", response_model=None)
async def learn_pet_skill_from_book(
    pet_id: int,
    payload: PetSkillLearnBookRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """消耗技能书学会技能（校验 scope）。"""
    character = await gate.require_character(current_user)
    data = await service.learn_skill_from_book(
        character,
        pet_id,
        book_id=payload.book_id,
    )
    return success(data)


@router.patch("/{pet_id}", response_model=None)
async def patch_pet(
    pet_id: int,
    payload: PetPatchRequest,
    service: PetService = Depends(get_pet_service),
    gate: PlayGate = Depends(get_play_gate),
    current_user: User = Depends(get_current_user),
) -> dict:
    """更新灵宠昵称或布阵偏好。"""
    character = await gate.require_character(current_user)
    data = await service.patch_pet(
        character,
        pet_id,
        nickname=payload.nickname,
        is_deploy_preferred=payload.is_deploy_preferred,
    )
    return success(data)

File: frontend/src/api/pets.ts
/**
 * M4/N4 ?? API?
 */
import { http } from './http'
import { envelopeFromAxiosError } from './envelope'
import type { ApiResponse } from '../types/api'
import type {
  PetCatalogPayload,
  PetDuelState,
  PetHatchState,
  PetCaptureResult,
  PetEncounterPublic,
  PetExplorePreview,
  PetPublic,
} from '../types/pets'

/** GET /pets */
export async function fetchPets(): Promise<
  ApiResponse<{ pets: PetPublic[]; hold_cap?: number }>
> {
  try {
    const response = await http.get<ApiResponse<{ pets: PetPublic[]; hold_cap?: number }>>(
      '/pets',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{ pets: PetPublic[]; hold_cap?: number }>(error)
  }
}

/** GET /pets/catalog */
export async function fetchPetCatalog(): Promise<ApiResponse<PetCatalogPayload>> {
  try {
    const response = await http.get<ApiResponse<PetCatalogPayload>>('/pets/catalog')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<PetCatalogPayload>(error)
  }
}

/** POST /pets/capture_test ? ?? species_id ???????? */
export async function captureTestPet(body?: {
  species_id?: string | null
}): Promise<
  ApiResponse<{ id: number; species_id: string; grade?: number; pet?: PetPublic }>
> {
  try {
    const payload =
      body?.species_id != null && String(body.species_id).trim() !== ''
        ? { species_id: body.species_id }
        : { species_id: null }
    const response = await http.post<
      ApiResponse<{ id: number; species_id: string; grade?: number; pet?: PetPublic }>
    >('/pets/capture_test', payload)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      species_id: string
      grade?: number
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/upgrade */
export async function upgradePet(
  petId: number,
): Promise<ApiResponse<{ id: number; level: number; stats?: PetPublic['stats']; pet?: PetPublic }>> {
  try {
    const response = await http.post<
      ApiResponse<{ id: number; level: number; stats?: PetPublic['stats']; pet?: PetPublic }>
    >(`/pets/${petId}/upgrade`)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      level: number
      stats?: PetPublic['stats']
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/grade-up */
export async function gradeUpPet(
  petId: number,
): Promise<
  ApiResponse<{
    id: number
    grade: number
    spirit_stones_spent: number
    spirit_stones: number
    pet?: PetPublic
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        id: number
        grade: number
        spirit_stones_spent: number
        spirit_stones: number
        pet?: PetPublic
      }>
    >(`/pets/${petId}/grade-up`)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      grade: number
      spirit_stones_spent: number
      spirit_stones: number
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/affix/reroll-value */
export async function rerollPetAffixValue(
  petId: number,
  slotIndex: number,
): Promise<
  ApiResponse<{
    id: number
    slot_index: number
    spirit_stones_spent: number
    spirit_stones: number
    pet?: PetPublic
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        id: number
        slot_index: number
        spirit_stones_spent: number
        spirit_stones: number
        pet?: PetPublic
      }>
    >(`/pets/${petId}/affix/reroll-value`, { slot_index: slotIndex })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      slot_index: number
      spirit_stones_spent: number
      spirit_stones: number
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/feed (PET-D04) */
export async function feedPet(
  petId: number,
  itemId: string,
  quantity = 1,
): Promise<
  ApiResponse<{
    id: number
    item_id: string
    quantity: number
    times_fed?: number
    total_used?: number
    total_cap?: number
    pet?: PetPublic
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        id: number
        item_id: string
        quantity: number
        times_fed?: number
        total_used?: number
        total_cap?: number
        pet?: PetPublic
      }>
    >(`/pets/${petId}/feed`, { item_id: itemId, quantity })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      item_id: string
      quantity: number
      times_fed?: number
      total_used?: number
      total_cap?: number
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/sect/affix/reroll-type (PET-D06) */
export async function rerollPetAffixType(
  petId: number,
  slotIndex: number,
): Promise<
  ApiResponse<{
    id: number
    slot_index: number
    spirit_stones_spent: number
    spirit_stones: number
    type_reroll_count?: number
    pet?: PetPublic
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        id: number
        slot_index: number
        spirit_stones_spent: number
        spirit_stones: number
        type_reroll_count?: number
        pet?: PetPublic
      }>
    >('/pets/sect/affix/reroll-type', { pet_id: petId, slot_index: slotIndex })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      slot_index: number
      spirit_stones_spent: number
      spirit_stones: number
      type_reroll_count?: number
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/skills/equip */
export async function equipPetSkills(
  petId: number,
  equipped: Array<string | null>,
): Promise<ApiResponse<{ id: number; equipped_ids?: Array<string | null>; pet?: PetPublic }>> {
  try {
    const response = await http.post<
      ApiResponse<{ id: number; equipped_ids?: Array<string | null>; pet?: PetPublic }>
    >(`/pets/${petId}/skills/equip`, { equipped })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      equipped_ids?: Array<string | null>
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/skills/learn */
export async function learnPetSkillFromPool(
  petId: number,
  skillId: string,
): Promise<ApiResponse<{ id: number; learned_skill_id?: string; pet?: PetPublic }>> {
  try {
    const response = await http.post<
      ApiResponse<{ id: number; learned_skill_id?: string; pet?: PetPublic }>
    >(`/pets/${petId}/skills/learn`, { skill_id: skillId })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      learned_skill_id?: string
      pet?: PetPublic
    }>(error)
  }
}

/** POST /pets/{id}/skills/learn_book */
export async function learnPetSkillFromBook(
  petId: number,
  bookId: string,
): Promise<
  ApiResponse<{ id: number; book_id?: string; learned_skill_id?: string; pet?: PetPublic }>
> {
  try {
    const response = await http.post<
      ApiResponse<{ id: number; book_id?: string; learned_skill_id?: string; pet?: PetPublic }>
    >(`/pets/${petId}/skills/learn_book`, { book_id: bookId })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      book_id?: string
      learned_skill_id?: string
      pet?: PetPublic
    }>(error)
  }
}

/** PATCH /pets/{id} */
export async function patchPet(
  petId: number,
  body: { nickname?: string; is_deploy_preferred?: boolean },
): Promise<
  ApiResponse<{ id: number; nickname?: string | null; is_deploy_preferred?: boolean }>
> {
  try {
    const response = await http.patch<
      ApiResponse<{ id: number; nickname?: string | null; is_deploy_preferred?: boolean }>
    >(`/pets/${petId}`, body)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      id: number
      nickname?: string | null
      is_deploy_preferred?: boolean
    }>(error)
  }
}

/** POST /pets/duel/npc/start */
export async function startPetDuelNpc(body: {
  pet_id: number
  npc_id?: string | null
  seed?: number | null
}): Promise<
  ApiResponse<{ duel_id: string; npc_id: string; seed: number; state: PetDuelState }>
> {
  try {
    const response = await http.post<
      ApiResponse<{ duel_id: string; npc_id: string; seed: number; state: PetDuelState }>
    >('/pets/duel/npc/start', body)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      duel_id: string
      npc_id: string
      seed: number
      state: PetDuelState
    }>(error)
  }
}

/** POST /pets/duel/{id}/turn */
export async function turnPetDuel(
  duelId: string,
  skillId: string | null,
): Promise<
  ApiResponse<{
    state: PetDuelState
    turn_events: Array<Record<string, unknown>>
    finished: boolean
    winner: string | null
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        state: PetDuelState
        turn_events: Array<Record<string, unknown>>
        finished: boolean
        winner: string | null
      }>
    >(`/pets/duel/${duelId}/turn`, { skill_id: skillId })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      state: PetDuelState
      turn_events: Array<Record<string, unknown>>
      finished: boolean
      winner: string | null
    }>(error)
  }
}

/** POST /pets/duel/npc/auto */
export async function autoPetDuelNpc(body: {
  pet_id: number
  npc_id?: string | null
  seed?: number | null
}): Promise<
  ApiResponse<{
    report: {
      winner: string | null
      rounds: number
      seed: number
      events: Array<Record<string, unknown>>
    }
    state: PetDuelState
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        report: {
          winner: string | null
          rounds: number
          seed: number
          events: Array<Record<string, unknown>>
        }
        state: PetDuelState
      }>
    >('/pets/duel/npc/auto', body)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      report: {
        winner: string | null
        rounds: number
        seed: number
        events: Array<Record<string, unknown>>
      }
      state: PetDuelState
    }>(error)
  }
}

/** GET /pets/hatch (N5) */
export async function fetchPetHatch(): Promise<ApiResponse<PetHatchState>> {
  try {
    const response = await http.get<ApiResponse<PetHatchState>>('/pets/hatch')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<PetHatchState>(error)
  }
}

/** POST /pets/hatch/start */
export async function startPetHatch(
  eggItemId: string,
): Promise<ApiResponse<{ job: PetHatchState['jobs'][number]; spirit_stones?: number }>> {
  try {
    const response = await http.post<
      ApiResponse<{ job: PetHatchState['jobs'][number]; spirit_stones?: number }>
    >('/pets/hatch/start', { egg_item_id: eggItemId })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      job: PetHatchState['jobs'][number]
      spirit_stones?: number
    }>(error)
  }
}

/** POST /pets/hatch/{id}/claim */
export async function claimPetHatch(
  jobId: number,
): Promise<ApiResponse<{ pet?: PetPublic; id?: number; job?: PetHatchState['jobs'][number] }>> {
  try {
    const response = await http.post<
      ApiResponse<{ pet?: PetPublic; id?: number; job?: PetHatchState['jobs'][number] }>
    >(`/pets/hatch/${jobId}/claim`)
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      pet?: PetPublic
      id?: number
      job?: PetHatchState['jobs'][number]
    }>(error)
  }
}

/** GET /pets/explore/preview (M4-D04c) */
export async function fetchPetExplorePreview(
  regionId = 'default',
): Promise<ApiResponse<PetExplorePreview>> {
  try {
    const response = await http.get<ApiResponse<PetExplorePreview>>('/pets/explore/preview', {
      params: { region_id: regionId },
    })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<PetExplorePreview>(error)
  }
}

/** POST /pets/explore/encounter */
export async function explorePetEncounter(body?: {
  region_id?: string
  seed?: number | null
}): Promise<ApiResponse<PetEncounterPublic>> {
  try {
    const response = await http.post<ApiResponse<PetEncounterPublic>>('/pets/explore/encounter', {
      region_id: body?.region_id ?? 'default',
      seed: body?.seed ?? null,
    })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<PetEncounterPublic>(error)
  }
}

/** POST /pets/explore/capture */
export async function explorePetCapture(body: {
  encounter_id: string
  seed?: number | null
}): Promise<ApiResponse<PetCaptureResult>> {
  try {
    const response = await http.post<ApiResponse<PetCaptureResult>>('/pets/explore/capture', {
      encounter_id: body.encounter_id,
      seed: body.seed ?? null,
    })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<PetCaptureResult>(error)
  }
}

/** POST /pets/explore/auto */
export async function explorePetAuto(body?: {
  region_id?: string
  seed?: number | null
}): Promise<
  ApiResponse<{
    rolls: PetEncounterPublic[]
    capture: PetCaptureResult | null
    seed: number
    auto: boolean
  }>
> {
  try {
    const response = await http.post<
      ApiResponse<{
        rolls: PetEncounterPublic[]
        capture: PetCaptureResult | null
        seed: number
        auto: boolean
      }>
    >('/pets/explore/auto', {
      region_id: body?.region_id ?? 'default',
      seed: body?.seed ?? null,
    })
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{
      rolls: PetEncounterPublic[]
      capture: PetCaptureResult | null
      seed: number
      auto: boolean
    }>(error)
  }
}
